refactor(linter): Update various rules to use DefaultRuleConfig and fix some minor issues.

connorshea opened the pull request:
Update a handful of rules to use the DefaultRuleConfig pattern.


- `react/button-has-type`
- `react/jsx-no-target-blank`
- `react/style-prop-object`
- `vue/define-props-destructuring`


For `react/style-prop-object`, this _was_ sorting the `allow` vector to enable binary search, but I'd be surprised if the performance of this matters enough to warrant it. And it wasn't being done consistently anyway, the most common case (JSX elements) wasn't even using a binary search.

This also fixes the rule docs for `vue/define-props-destructuring`, which were previously broken.

Now they look like this:

```md
## Configuration

This rule accepts a configuration object with the following properties:

### destructure

type: `"always" | "never"`

default: `"always"`

Require or prohibit destructuring.

#### `"always"`

Requires destructuring when using `defineProps` and warns against using `withDefaults` with destructuring

#### `"never"`

Requires using a variable to store props and prohibits destructuring
```

camc314 commented:
### Merge activity

* **Dec 14, 1:41 PM UTC**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **Dec 14, 1:41 PM UTC**: `camc314` added this pull request to the [Graphite merge queue](https://app.graphite.com/merges?org=oxc-project&repo=oxc).

camc314 commented on crates/oxc_linter/src/rules/vue/define_props_destructuring.rs:
```diff
@@ -25,15 +25,19 @@ fn avoid_with_defaults_diagnostic(span: Span) -> OxcDiagnostic {
+    /// Require or prohibit destructuring.
     destructure: Destructure,
 }
 
```
I guess `[serde(rename_all = "kebab-case")]` is unnceessary since they're single strings?

connorshea committed:
refactor(linter): Update various rules to use DefaultRuleConfig and fix some minor issues. (#16816)

Update a handful of rules to use the DefaultRuleConfig pattern.

- `react/button-has-type`
- `react/jsx-no-target-blank`
- `react/style-prop-object`
- `vue/define-props-destructuring`

For `react/style-prop-object`, this _was_ sorting the `allow` vector to enable binary search, but I'd be surprised if the performance of this matters enough to warrant it. And it wasn't being done consistently anyway, the most common case (JSX elements) wasn't even using a binary search.

This also fixes the rule docs for `vue/define-props-destructuring`, which were previously broken.

Now they look like this:

```md
## Configuration

This rule accepts a configuration object with the following properties:

### destructure

type: `"always" | "never"`

default: `"always"`

Require or prohibit destructuring.

#### `"always"`

Requires destructuring when using `defineProps` and warns against using `withDefaults` with destructuring

#### `"never"`

Requires using a variable to store props and prohibits destructuring
```
